### PR TITLE
added `KafkaDsn` to `networks.py` and add default ports for `HttpUrl` 

### PR DESCRIPTION
**Pull request by MihanixA**
❤️ pydantic

## Change Summary

- Added `KafkaDsn` class e.g. already existing `RedisDsn`  :fire:
- Added `default_parts` member and `apply_default_parts` method to the `AnyUrl` class

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

**Comment by davidolrik** [comment]
I think this will also allow a fix for my issue #2450, where I want default ports on http urls.

Will a stringification include the defaults? or will the original string be preserved?

**Comment by MihanixA** [comment]
> I think this will also allow a fix for my issue #2450, where I want default ports on http urls.
> 
> Will a stringification include the defaults? or will the original string be preserved?

An example of its behaviour is shown in the test for `KafkaDsn` model. Default host and port are included to the str version. 

https://github.com/samuelcolvin/pydantic/pull/2447/files#diff-f953eed7c7855cc439287af4864f147a1ed9c36995d18f335b57b96e3d7653b3R385

**Comment by MihanixA** [comment]
@davidolrik I could commit this solution and link your issue to this ticket. What do you think?

**Comment by MihanixA** [comment]
I've also added default `443` or `80` ports to `HttpUrl` model which are accessible but not displayed in the stringified version to this PR. 
@davidolrik @samuelcolvin

**Comment by MihanixA** [comment]
I've also added default 443 or 80 ports to HttpUrl model which are accessible but not displayed in the stringified version to this PR.
@davidolrik @samuelcolvin

**Comment by davidolrik** [comment]
Just what I was thinking! 😁

Maybe also add test for http urls with ports included which is also one of the defaults?

```python
m = Model(v='https://www.example.com:443')
assert m.v.port == '443'
assert m.v == 'https://www.example.com:443'
```

**Comment by davidolrik** [comment]
I was just playing around with your branch, and I found a bug:

```python
m = Model(v='https://www.example.com:1234')
assert m.v.port == '1234'
assert m.v == 'https://www.example.com:1234'
```

Here the port is also set to 443:

```
>       assert m.v.port == '1234'
E       AssertionError: assert '443' == '1234'
E         - 1234
E         + 443
```

**Comment by MihanixA** [comment]
I didn't notice any valid test in `test_networks.py` with port except 443 and 80 and I decided for a moment that behaviour with ports like 1234 is unsupported. 
I had already been working on fix by the time I read your message though 
But thank you any way! :)

**Comment by MihanixA** [comment]
@davidolrik added tests as you requested

**Comment by MihanixA** [comment]
#2450

**Comment by MihanixA** [comment]
@samuelcolvin ready for review

**Comment by MihanixA** [comment]
@PrettyWood I'd appreciate review and/or some feedback 🛩️ 😃

File: pydantic/networks.py [hunk 1]
```diff
@@ -109,6 +110,8 @@ class AnyUrl(str):
     allowed_schemes: Optional[Set[str]] = None
     tld_required: bool = False
     user_required: bool = False
+    default_parts: Optional[Set[Tuple[str, str]]] = None
```
Comment (PrettyWood) [inline pydantic/networks.py:113]: Why not a dictionary? or even a method like
```py
class AnyHttpUrl(AnyUrl):
    allowed_schemes = {'http', 'https'}

    @staticmethod
    def get_default_parts(parts: Dict[str, str]) -> Dict[str, str]:
        return {
            'port': '443' if parts['scheme'] == 'https' else '80',
        }

class HttpUrl(AnyHttpUrl):
    tld_required = True
    # https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers
    max_length = 2083

class RedisDsn(AnyUrl):
    allowed_schemes = {'redis', 'rediss'}

    @staticmethod
    def get_default_parts(parts: Dict[str, str]) -> Dict[str, str]:
        return {
            'domain': 'localhost' if not (parts['ipv4'] or parts['ipv6']) else '',
            'port': '6379',
            'path': '/0',
        }
```

File: pydantic/networks.py [hunk 2]
```diff
@@ -109,6 +110,8 @@ class AnyUrl(str):
...
+    display_port: bool = True
```
Comment (PrettyWood) [inline pydantic/networks.py:114]: Hmmm I think this issue doesn't concern only the port. The question is "when a value is set by magic, should the string be changed". As a matter of fact I think you don't need this. What matters is `rebuild`
```py
host, tld, host_type, rebuild = cls.validate_host(parts)
```
which, when set to `True`, will rebuild the url

File: pydantic/networks.py [hunk 3]
```diff
@@ -281,6 +285,14 @@ def validate_host(cls, parts: Dict[str, str]) -> Tuple[str, Optional[str], str,
 
         return host, tld, host_type, rebuild  # type: ignore
 
+    @classmethod
+    def apply_default_parts(cls, parts: Dict[str, str]) -> Dict[str, str]:
```
Comment (PrettyWood) [inline pydantic/networks.py:289]: So here you could have something like
```py
    @classmethod
    def apply_default_parts(cls, parts: Dict[str, str]) -> Dict[str, str]:
        for part, value in cls.get_default_parts(parts).items():
            if parts[part] is None:
                parts[part] = value
        return parts

    @staticmethod
    def get_default_parts(parts: Dict[str, str]) -> Dict[str, str]:
        return {}
```

**Comment by MihanixA** [comment]
I couldn't make this work with default port in `HttpUrl` for now. I really tried though
So I guess I am going to commit fixes for the PR issues and try to solve `HttpUrl` problem in a separate PR..

**Comment by MihanixA** [comment]
I added some `# noqa` (pls tell if you want them removed) 
And also made `HttpUrl` inherit `AnyHttpUrl` instead of `AnyUrl`

File: pydantic/networks.py (hunk 2 above)
Comment (MihanixA) [reply pydantic/networks.py]: I didn't find a good way to implement this..

File: pydantic/networks.py (hunk 1 above)
Comment (MihanixA) [reply pydantic/networks.py]: Fixed

Comment (MihanixA) [reply pydantic/networks.py]: I guess I wanted to use set to have FrozenSet for typing somewhere

File: pydantic/networks.py (hunk 2 above)
Comment (MihanixA) [reply pydantic/networks.py]: I had to reverted the "default http port" behaviour

File: changes/2447-MihanixA.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+added `KafkaDsn` to the `pydantic/networks.py`
```
[the remark this answers is not in the document]
Comment (MihanixA) [reply changes/2447-MihanixA.md]: fixed

File: pydantic/networks.py (hunk 3 above)
Comment (MihanixA) [reply pydantic/networks.py:291]: Done!

**Comment by MihanixA** [comment]
I noticed that CI test "test FastAPI" failed. I run through an error message and I not sure if it's caused by my changes..

**Comment by MihanixA** [comment]
@PrettyWood 🙃 🎉

File: pydantic/networks.py [hunk 5]
```diff
@@ -114,7 +115,7 @@ class AnyUrl(str):
 
     @no_type_check
     def __new__(cls, url: Optional[str], **kwargs) -> object:
-        return str.__new__(cls, cls.build(**kwargs) if url is None else url)
+        return str.__new__(cls, cls.build(**kwargs) if url is None else url)  # noqa
```
Comment (PrettyWood) [inline pydantic/networks.py:118]: I would remove the `# noqa`. I understand some IDEs like Pycharm will display some kind of warning but as long as `flake8` says nothing we're good

**Comment by PrettyWood** [comment]
Thanks! Don't worry about the CI it's due to the new sqlalchemy 1.4 version that is not pinned nor supported by `encoded/databases`. I opened https://github.com/encode/databases/pull/299

**Comment by MihanixA** [comment]
Thank you for feedback!

> For the default port don't we want to add
> 
> ```python
>     @staticmethod
>     def get_default_parts(parts: 'Parts') -> 'PartialParts':
>         return {
>             'port': '80' if parts['scheme'] == 'http' else '443',
>         }
> ```

Maybe use only `'PartialParts'` without `'Parts'`? By the time `get_default_parts` is called, `parts` argument doesn't comply with described `TypedDict`. Besides this move would allow to avoid code duplication.

**Comment by MihanixA** [comment]
@PrettyWood 👻 🥳

**Comment by MihanixA** [comment]
> I'd like to continue to work on that issue in a separate PR when the situation about default parts is finalized)

Nevermind :)

**Comment by PrettyWood** [comment]
Can you rebase please ?

**Comment by MihanixA** [comment]
@PrettyWood done as you requested.

All tests passed according to CI. 
Though when I run tests locally, I get this:
```
========================================================================================= short test summary info ==========================================================================================
FAILED tests/test_generics.py::test_replace_types - IndexError: tuple index out of range
FAILED tests/test_schema.py::test_unenforced_constraints_schema[kwargs9-type_9] - IndexError: tuple index out of range

Results (57.85s):
    2045 passed
       2 failed
         - tests/test_generics.py:787 test_replace_types
         - tests/test_schema.py:1442 test_unenforced_constraints_schema[kwargs9-type_9]
       6 skipped
make: *** [test] Error 1
```

Also codecov is less then 100% and I am not sure why :c

**Comment by PrettyWood** [comment]
You have all the commits from master that unrelated. Please rebase on the upstream version and push force to have a clean change diff. And we'll see after for the coverage

**Comment by MihanixA** [comment]
Looks like I broke everything in this branch(
I am going to create close this branch and create a new one with clear change history
OK? @PrettyWood

**Comment by PrettyWood** [comment]
Lol no you didn't ;) it's just your rebase that was done on your local version of master and not the upstream one. I can rewrite your history if you need help

**Comment by MihanixA** [comment]
> Lol no you didn't ;) it's just your rebase that was done on your local version of master and not the upstream one. I can rewrite your history if you need help

I'd appreciate that!

**Comment by MihanixA** [comment]
All tests passed. That is great! 
The only thing that remains is to push merge button 👻

**Comment by MihanixA** [comment]
@PrettyWood

**Comment by PrettyWood** [comment]
Samuel merges the PRs. And a 1.8.2 may be released before 1.9.

**Comment by MihanixA** [comment]
> Samuel merges the PRs. And a 1.8.2 may be released before 1.9.

@samuelcolvin ok, thanks! Did you forget to click 'Approve' button or there is something I should fix?

**Review by PrettyWood — APPROVED** [review]

**Comment by PrettyWood** [comment]
Thank you!

**Pull request merged**